concourse/scripts/entry.sh

```diff
@@ -0,0 +1,136 @@
+    version=$(awk -F " *= *" '$1 == "VERSION_ID" { print $2 }' /etc/os-release | tr -d \")
+  else
+    echo "Could not determine operating system type" >/dev/stderr
+    exit 1
+  fi
+  echo "${name}${version}"
+}
+
+OS_NAME=$(_determine_os)
+
+# Global ENV defines
+# /tmp/build/xxxxx. it should not be used in normal conditions. Use /home/gpadmin instead.
+# Everything has been linked there.
+export CONCOURSE_WORK_DIR=${PWD}
+
+# Create gpadmin user and chown all files in the PWD. All files will be linked to /home/gpadmin.
+# All of our work should be started from there.
+setup_gpadmin() {
+    # If the gpadmin exist, quit
+    if grep -c '^gpadmin:' /etc/passwd; then
+        return
+    fi
+
+    # If the image has sshd, then we call gpdb's setup_gpadmin_user.sh to create the gpadmin user
+    # and setup the ssh.
+    # Otherwise, create the gpadmin user only.
+    if [ -f /etc/ssh/sshd_config ]; then
+        local gpdb_concourse_dir="${CONCOURSE_WORK_DIR}/gpdb_src/concourse/scripts"
+        "${gpdb_concourse_dir}/setup_gpadmin_user.bash"
+    else
+        local test_os=$(_determine_os)
+        # Below is copied from setup_gpadmin_user.bash
+        groupadd supergroup
+        case "$test_os" in
+            rhel*)
+                /usr/sbin/useradd -U -G supergroup,tty -m gpadmin
+                ;;
+            ubuntu*)
+                /usr/sbin/useradd -U -G supergroup,tty -m -s /bin/bash gpadmin
+                ;;
+            sles*)
+                # create a default group gpadmin, and add user gpadmin to group gapdmin, supergroup,
+                # tty
+                /usr/sbin/useradd -U -G supergroup,tty -m gpadmin
+                ;;
+            photon*)
+                /usr/sbin/useradd -U -G supergroup,tty,root -m gpadmin
+                ;;
+            *) echo "Unknown OS: $test_os"; exit 1 ;;
+        esac
+        echo -e "password\npassword" | passwd gpadmin
+    fi
+
+    chown -R gpadmin:gpadmin /tmp/build
+    ln -s "${CONCOURSE_WORK_DIR}"/* /home/gpadmin
+}
+
+# Extract gpdb binary
+function install_gpdb() {
+    [ ! -d /usr/local/greenplum-db-devel ] && mkdir -p /usr/local/greenplum-db-devel
+    tar -xzf "${CONCOURSE_WORK_DIR}"/bin_gpdb/*.tar.gz -C /usr/local/greenplum-db-devel
+    chown -R gpadmin:gpadmin /usr/local/greenplum-db-devel
+}
+
+function setup_gpadmin_bashrc() {
+    {
+        echo "source /usr/local/greenplum-db-devel/greenplum_path.sh"
+        echo "export OS_NAME=${OS_NAME}"
+    } >> /home/gpadmin/.bashrc
+}
+
+# Setup common environment
+setup_gpadmin
+install_gpdb
+setup_gpadmin_bashrc
+
+# Do the special setup with root permission for the each task, then run the real task script with
+# gpadmin. bashrc won't be read by 'su', it needs to be sourced explicitly.
+case "$1" in
+    build)
+        echo "build: OS_NAME=${OS_NAME}"
+        su gpadmin -c \
+           "source /home/gpadmin/.bashrc &&\
+            . /home/gpadmin/postgresql-hll_src/concourse/scripts/build_postgresql-hll.sh"
```

Comment:
should not have `.`

Reply:
I set the variable in the command passed to `su`, so I have to use source. Executing the script directly on ubuntu still does not work. The script can not get the value of the variable `OS_NAME`.

Reply:
The root cause of the problem is that Ubuntu has a default `.bashrc` file by default which can not be `sourced` by command or script. It's solved by overwriting the `.bashrc` file in this commit: 1df873ccf33a222ece0046556c1d43a8cd1c8321